(self.webpackChunk_N_E=self.webpackChunk_N_E||[]).push([[5155],{9480:function(n,e,a){(window.__NEXT_P=window.__NEXT_P||[]).push(["/tutorials/deepchem-antibodytutorial-simplified",function(){return a(2052)}])},2052:function(n,e,a){"use strict";a.r(e),a.d(e,{default:function(){return c}});var s=a(5893),t=a(1618),i=a(6485),o={html:'<main>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h1>\n      <strong>\n       An Introduction to Antibody Design Using Protein Language Models\n      </strong>\n      <a class="anchor-link" href="#An-Introduction-to-Antibody-Design-Using-Protein-Language-Models">\n       \xb6\n      </a>\n     </h1>\n     <h4 id="By-Dhuvi-Karthikeyan-and-Aaron-Menezes">\n      By Dhuvi Karthikeyan and Aaron Menezes\n      <a class="anchor-link" href="#By-Dhuvi-Karthikeyan-and-Aaron-Menezes">\n       \xb6\n      </a>\n     </h4>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <p>\n      This DeepChem tutorial is designed to serve as a brief primer for antibody design via protein language models. Antibodies are immune proteins also known as immunoglobulins that are naturally produced in the body and bind/inactivate viruses and other pathogens. They are a valuable therapeutic and save lives in immune checkpoint inhibitors for cancer and as neutralizing antibodies for acute viral infection. In addition, they are well known outside the hospital walls for their ability to bind and stick to arbitrary molecular targets, a useful feature in the basic sciences and industrial biochemical facilities.\n     </p>\n     <p>\n      This tutorial aims to provide a quick overview of key immunology concepts needed to understand antibody structure and function in the broader context of the immune system. We make some assumptions with familiarity with large language models. Take a look at our other\n      <a href="https://colab.research.google.com/drive/13eXPgZpzTOL3c_S7OM7uR6btP8ZWw7zn">\n       tuorial\n      </a>\n      . For the sake of brevity we provide links on non-essential topics that point to external sources wherever possible. Follow along to learn more about the immune system, and protein language models for guided Ab design.\n     </p>\n     <p>\n      <strong>\n       Note:\n      </strong>\n      This tutorial is loosely based on the 2023 Nature Biotechnology paper titled "Efficient evolution of human antibodies from general protein language models"\n      <a href="https://www.nature.com/articles/s41587-023-01763-2#Sec37">\n       [1]\n      </a>\n      by Hie et al. We thank the authors for making their methods and data available and accessible.\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h2>\n      1. Immunology 101\n      <a class="anchor-link" href="#1.-Immunology-101">\n       \xb6\n      </a>\n     </h2>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h3 id="1.1-What-is-the-immune-system?">\n      1.1 What is the immune system?\n      <a class="anchor-link" href="#1.1-What-is-the-immune-system?">\n       \xb6\n      </a>\n     </h3>\n     <p>\n      Like other body systems such as the digestive system or the cardiovascular system, the immune system is fundamentally a collection of specialized cells  that operate together to accomplish a specific\n      <a href="https://en.wikipedia.org//wiki/Homeostasis">\n       homeostatic\n      </a>\n      function. The immune system is responsible for the protection of our body\'s vast resources (lipids, carbodhydrates, enzymes, proteins), especially against opportunistic threats of the outside world, such as viruses and bacteria. A very simple yet powerful component of the immune system is the skin, which keeps what\'s out, out and what\'s in, in. However, what happens when we get a cut/scrape or when we walk past someone who\'s coughing and accidentally breathe in the droplets? Thankfully, it turns out the immune system has a whole army of highly specialized white blood cells patrolling our blood and lymph nodes, ready to launch a multi-layered response against these and the countless other cases where foreign objects may enter our bodies. Through these white bloods cells, the immune system is able to accomplish its primary function,\n      <strong>\n       self-nonself discrimination\n      </strong>\n      , the accurate identification of molecules that are sufficiently dissimilar than those derived from a healthy person.\n     </p>\n     <p>\n      If you would like to learn more about the complex problem of self non-self discrimination, and appreciate the theory behind the immune system\'s organization and function, we recommend checking out the following works:\n     </p>\n     <ul>\n      <li>\n       A Theory of Self-Nonself Discrimination\n       <a href="https://www.science.org/doi/10.1126/science.169.3950.1042">\n        [2]\n       </a>\n      </li>\n      <li>\n       The common sense of the self-nonself discrimination, 2005\n       <a href="https://link.springer.com/article/10.1007/s00281-005-0199-1">\n        [3]\n       </a>\n      </li>\n      <li>\n       Conceptual aspects of self and nonself discrimination, 2011\n       <a href="https://www.tandfonline.com/doi/epdf/10.4161/self.2.1.15094?needAccess=true">\n        [4]\n       </a>\n      </li>\n      <li>\n       Self-Nonself Discrimination due to Immunological Nonlinearities: the Analysis of a Series of Models by Numerical Methods, 1987\n       <a href="https://academic.oup.com/imammb/article/4/1/1/875656">\n        [5]\n       </a>\n      </li>\n      <li>\n       A biological context for the self-nonself discrimination and the regulation of effector class by the immune system, 2005\n       <a href="https://link.springer.com/article/10.1385/IR:31:2:133">\n        [6]\n       </a>\n      </li>\n     </ul>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h3 id="1.2-Innate-vs.-Adaptive-Immune-System">\n      <strong>\n       1.2 Innate vs. Adaptive Immune System\n      </strong>\n      <a class="anchor-link" href="#1.2-Innate-vs.-Adaptive-Immune-System">\n       \xb6\n      </a>\n     </h3>\n     <p>\n      Remarkably, these\n      <a href="https://en.wikipedia.org/wiki/White_blood_cell">\n       white blood cells\n      </a>\n      operate independently, adding and subtracting from the local context (tissue microenvironment), orchestrating cohesive responses in a completely decentralized manner. Individual white blood cells interact with their environments through general\n      <a href="https://en.wikipedia.org/wiki/Cytokine">\n       cytokine\n      </a>\n      and\n      <a href="https://en.wikipedia.org/wiki/Chemokine">\n       chemokine\n      </a>\n      surface receptors that facilitate gradient sensing of the local environment. This is the primary axis by which immune cells are able to home to sites of infection/wound healing and send pro-inflammatory or anti-inflammatory signals to neighboring cells. In addition, nearly all immune cells require some level of activation by means of receptor:ligand stimulation by non-self signatures. How specific these receptors are helps highlight the delineation between the two arms of the immune system: the innate and adaptive immune systems.\n     </p>\n     <h4 id="1.2.1-The-Innate-Immune-System">\n      1.2.1 The Innate Immune System\n      <a class="anchor-link" href="#1.2.1-The-Innate-Immune-System">\n       \xb6\n      </a>\n     </h4>\n     <p>\n      The\n      <a href="https://en.wikipedia.org/wiki/Innate_immune_system">\n       innate immune system\n      </a>\n      is the body\'s first line of defense, and encompasses different types of cells (and proteins) that recognize broadly non-self signals known as pathogen-associated or damage-associated molecular patterns (\n      <a href="https://en.wikipedia.org/wiki/Pathogen-associated_molecular_pattern">\n       PAMPs\n      </a>\n      or\n      <a href="https://en.wikipedia.org/wiki/Damage-associated_molecular_pattern">\n       DAMPs\n      </a>\n      ). The effector cells of the innate immune system are decorated with ~20 different kinds of pattern recognition receptors (PRRs) that recognize broad signals (single stranded DNA, LPS, other signals of pathogen activity). Being the first responders, the innate immune system handles containment, engulfing the offending signals and otherwise blocking it off, and tries to destroy the threat. Lastly, these cells simultaneously sound the alarm, putting the local area into a heightened state of threat detection and infection prevention, such as spiking a fever and initiating swelling to draw in more immune cells.\n     </p>\n     <h4 id="1.2.2-The-Adaptive-Immune-System">\n      1.2.2 The Adaptive Immune System\n      <a class="anchor-link" href="#1.2.2-The-Adaptive-Immune-System">\n       \xb6\n      </a>\n     </h4>\n     <p>\n      The\n      <a href="https://en.wikipedia.org/wiki/Adaptive_immune_system">\n       adaptive immune system\n      </a>\n      on the other hand, is slow to respond and is often brought in by innate cell activation. Instead of broad PRRs, it relies on tens of millions of\n      <a href="https://en.wikipedia.org/wiki/Somatic_recombination">\n       somatically rearranged\n      </a>\n      and stochastically generated highly specific adaptive immune receptors (AIRs) whose shape complementarity allows them to identify their complementary molecular patterns called epitopes. There are many more differences between the innate and adaptive immune systems, such as the latter\'s ability to develop a durable memory response, and those are introduced in greater detail here\n      <a href="https://www.ncbi.nlm.nih.gov/books/NBK27090/">\n       [7]\n      </a>\n      . It is important to note however, that it is through the independent and asynchronous operation of both the innate and adaptive immune system that we see the dynamics\n      <a href="https://www.nature.com/articles/nri700">\n       [8]\n      </a>\n      of threat detection, message passing, calling for reinforcements, homing of adaptive immune cells, and activation/expansion of these cells, resulting in complete clearance of the pathogen.Another key distinction between the innate and adaptive immune system is the formation of a robust memory response. Amazingly, evolution has steered this system to imprint past pathogen exposures so that upon re-exposure to the same signal, a small pool of memory cells is activated and antigen-specifc adaptive immune cells rapidly proliferatre and clear the antigen-source\n      <a href="https://asm.org/articles/2023/may/understanding-immunological-memory">\n       [9]\n      </a>\n      .\n     </p>\n     <pre class="overflow-x-scroll font-mono"><code class="language-algorithmic">algorithmic\nAdaptive Immune Algorithm\n\n1. Nascent progenitor cells undergoes somatic recombination to stochastically generate an Adaptive Immune Receptor (AIR).\n2. Self-selecting methods of ensuring self-tolerance remove cells that react too strongly to the self before being released into the blood.\n3. In periphery, naive (antigen-inexperienced) cells interact with antigens via AIRs in search of their cognate epitope.\n4. Upon recognition of a sufficiently strong epitope reaction, they divide rapidly and overwhelm the offending antigen with sheer numbers and specialized effector function.\n5. After the threat has cleared, this expanded population contracts as cells die without activation signalling, and a lasting pool of memory cells remains.\n6. Memory pool persists and is reactivated in an effector state upon antigen reintroduction.\n</code></pre>\n     <h3 id="1.2.3-Effector-Cells-of-the-Adaptive-Immune-System">\n      1.2.3 Effector Cells of the Adaptive Immune System\n      <a class="anchor-link" href="#1.2.3-Effector-Cells-of-the-Adaptive-Immune-System">\n       \xb6\n      </a>\n     </h3>\n     <p>\n      There are two major types of effector cells in the adaptive immune response:\n      <a href="https://en.wikipedia.org/wiki/T_cell">\n       T-cells\n      </a>\n      and\n      <a href="https://en.wikipedia.org/wiki/B_cell">\n       B-cells\n      </a>\n      . Both the T-cell and B-cell populations in the body are contextualized through their adaptive immune receptors, T-cell Receptors (TCRs) and B-cell Receptors (BCRs), respectively. We can think of both populations as repertoires of receptors, conferring protection against the threats recognized by the repertoire. T-cells help maitnain\n      <a href="https://en.wikipedia.org/wiki/Cell-mediated_immunity">\n       cellular immunity\n      </a>\n      by interrogating the intracellular component of our bodies\' by using their TCRs to scan recycled protein fragments presented at the cell surface (Read more about T-cell mediated immunity here\n      <a href="https://www.ncbi.nlm.nih.gov/books/NBK10762/">\n       [10]\n      </a>\n      ). B-cells, on the other hand, use their BCRs to survey the extracellular compartment and are tasked with upkeeping\n      <a href="https://en.wikipedia.org/wiki/Humoral_immunity">\n       humoral immunity\n      </a>\n      : neutralizing threats floating around in the blood and plasma. As effector cells of the adaptive immune response, both T-cell and B-cells are similar in their development and their operation as a unit, though the specifics of per cell function are quite different.\n     </p>\n     <p>\n      <strong>\n       Note\n      </strong>\n      : A helpful distinction between antigens and epitopes is that an antigen is something that broadly generates an immune response and can have multiple epitopes. Epitopes are specific molecular patterns that have a matching paratope (binding surface of an adaptive immune receptor).\n     </p>\n     <p>\n      <img alt="immune_system.png" src="https://www.creative-diagnostics.com/upload/image/innate%20and%20adaptive%20immunity2.png"/>\n     </p>\n     <p>\n      Image Source:\n      <a href="https://www.creative-diagnostics.com/innate-and-adaptive-immunity.htm">\n       Creative Diagnostics\n      </a>\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h3 id="1.3-B-Cells-and-Antibodies">\n      <strong>\n       1.3 B-Cells and Antibodies\n      </strong>\n      <a class="anchor-link" href="#1.3-B-Cells-and-Antibodies">\n       \xb6\n      </a>\n     </h3>\n     <p>\n      B-cells, or B-lymphocytes, get their name not from their origin in the bone marrow, but from their discovery in a particular organ of the chicken\n      <a href="https://www.sciencedirect.com/science/article/pii/S0032579119369536?via%3Dihub">\n       [11]\n      </a>\n      . These cells circulate in the bloodstream, equipped with unique B-cell receptors (BCRs) that allow them to recognize specific antigens, leading to their activation. This process often requires additional stimulation from helper T-cells which provide essential co-stimulatory signals, an additional layer of verification for non-selfness.\n     </p>\n     <p>\n      Over the course of the COVID pandemic, whether we wanted to or not were exposed to the concept of antibodies and learned of their association with some sort of protective capacity against SARS-COV-2. But what are they, and where do they originate from?\n     </p>\n     <p>\n      Antibodies (Abs) are typically represented as Y-shaped proteins that bind to their cognate epitope surfaces with high specificity and affinity, similar to how TCRs and BCRs bind to their epitopes. This is because antibodies are the soluble form of the B-cell receptor that is secreted into the blood upon B-cell activation in the presence of its cognate antigen. The secretion of large amounts of antibodies is the primary effector function of B-cells. Upon activation, a B-cell will divide, with the daughter cells inheriting the same BCR, and some of these cells will differentiate into plasma cells, which are the Ab factories capable of secreting thousands of Abs/min. This is especially useful upon antigen re-encounter where a large amount of antibodies are released by memory cells which neutralize the pathogen even before we develop the symptoms of infection (this is what most common vaccines are designed to do).\n     </p>\n     <p>\n      <img alt="b_cell_activation.png" src="https://media.beckman.com/-/media/stock-images/resource-center/images/b_cell_activation-2022-05.jpg?rev=4680b251612b4f0c9afc3280029d0cc7"/>\n     </p>\n     <p>\n      Image Source:\n      <a href="https://www.beckman.com/resources/cell-types/blood-cells/leukocytes/lymphocytes/b-cells">\n       Beckman\n      </a>\n     </p>\n     <p>\n      Neutralizing mechanisms of pathogenesis is only one way that antibody tagging is useful to immune defense. Antibody tagging plays a key role in a number of humoral immunity processes:\n     </p>\n     <ol>\n      <li>\n       <p>\n        <a href="https://en.wikipedia.org/wiki/Neutralizing_antibody">\n         Neutralization\n        </a>\n        : De-activation of pathogenic function by near complete coating of the functional component of pathogens or toxins by antibodies to inhibit interaction with host cells (i.e. and antibody that binds to the surface glycoproteins on SARS-COV2 now inhibit that virus particle\'s ability to enter cells expressing ACE2).\n       </p>\n      </li>\n      <li>\n       <p>\n        <a href="https://en.wikipedia.org/wiki/Antibody_opsonization">\n         Opsonization\n        </a>\n        : Partial coating of pathogens enhances rates phagocytosis and removal from the blood by cells of the innate immune system.\n       </p>\n      </li>\n      <li>\n       <p>\n        <a href="https://en.wikipedia.org/wiki/Agglutination_(biology">\n         Agglutination/Precipitation\n        </a>\n        ): Since antibodies have 2 arms (each arm of the Y), they can cross-link and form anitbody-antigens chains which can precipitate out of the plasma and increase their chances of being recognized as aberrant and cleared by phagocytes.\n       </p>\n      </li>\n      <li>\n       <p>\n        <a href="https://en.wikipedia.org/wiki/Complement_system">\n         Complement Activation\n        </a>\n        : The complement system is a collection of inactive proteins and protein precursors are self-amplifying on activation and help with multiple aspects of humoral immunity. Yet another function of antibodies is their role in initiating the complement cascade that ends in the lysis or phagocytosis of pathogens.\n       </p>\n      </li>\n     </ol>\n     <p>\n      <img alt="antibody_function.png" src="https://www.agproud.com/ext/resources/PD/images/stories/2021/03/11/031121-pd-weiland-fg1.jpg?t=1695665765&amp;width=610"/>\n     </p>\n     <p>\n      Image Source: The Immune System: Innate and Adaptive Body Defenses Figure 21.15 pulled from\n      <a href="https://www.agproud.com/articles/35884-understanding-vaccines">\n       [Source]\n      </a>\n     </p>\n     <p>\n      Given the importance of B-cell mediated immunity, as operationalized by the body\'s antibody repertoire, it\'s clear that the diversity of BCR clones plays a critical role in our ability to mount an effective response against a pathogen. The maintenance of a robust BCR repertoire highlights not only the complexity of the immune response but also underscores the potential for leveraging the modularity of this mechanism to introduce new clones for their extraordinary precision in therapeutics such as vaccine development.\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h3 id="1.4-Antibody-Sequence,-Structure,-and-Function">\n      <strong>\n       1.4 Antibody Sequence, Structure, and Function\n      </strong>\n      <a class="anchor-link" href="#1.4-Antibody-Sequence,-Structure,-and-Function">\n       \xb6\n      </a>\n     </h3>\n     <p>\n      The remarkable diversity of antibodies is achieved through somatic recombination, or gene rearrangement at the DNA level that occurs outside of meiosis. The AIR-specific somatic recombination is known as\n      <a href="https://en.wikipedia.org/wiki/VDJ_recombination">\n       V(D)J recombination\n      </a>\n      and generates both TCR and BCR diversity. During V(D)J recombination, a single gene per set is sampled from the set of variable (V), diversity (D), and joining (J) gene segments and randomly joined together with some baked in error (insertions) to create stable BCRs with unique antigen-binding sites. Additionally, B-cells have an additional process that further amplifies the diversity as well as the functional capacity of antigen-specific antibodies. This process is known as somatic hypermutation. When an activated B-cell divides, somatic hypermutation (SHM) introduces point mutations in the variable regions of BCR generating minor variants of each BCR. These daughter cells compete for survival signals mediated through antigen binding such that only the stronger binders survive.\n     </p>\n     <p>\n      Structurally, antibodies are composed of two identical light chains and two identical heavy chains, linked by disulfide bonds. Each chain contributes to the formation of the antigen-binding site, located in the variable regions. Within these regions, hypervariable loops known as complementarity determining regions (CDRs) dictate the specificity and affinity of the antibody-antigen interaction. This specificity is measured in terms of affinity using the dissociation constant (Kd), and the avidity (affinity over multi-valent binding sites, see\n      <a href="https://en.wikipedia.org/wiki/Immunoglobulin_M">\n       IgM\n      </a>\n      ,\n      <a href="https://en.wikipedia.org/wiki/Immunoglobulin_A">\n       IgA\n      </a>\n      ). The antibody molecule is divided into two main functional regions:\n     </p>\n     <ol>\n      <li>\n       Fab Region (Fragment, antigen-binding): Contains the variable regions of the light and heavy chains, responsible for antigen recognition and binding.\n      </li>\n      <li>\n       Fc Region (Fragment, crystallizable): Composed of the constant regions of the heavy chains, mediates interactions with innate immune cells and the complement system.\n      </li>\n     </ol>\n     <p>\n      <img alt="test_image" src="https://www.dianova.com/en/wp-content/uploads/sites/3/2021/01/Antikorperstruktur-IgG.jpg"/>\n     </p>\n     <p>\n      Image Source: Dianova:\n      <a href="https://www.dianova.com/en/faq/antibody-structure-what-is-a-secondary-antibody/">\n       Antibody Structure\n      </a>\n      \\\n     </p>\n     <p>\n      By harnessing the selection of evolutionary pressures during somatic hypermutation, the B-cell compartment uses a powerful method of further tuning the antibody specificity to have some of the highest affinity interactions in the known protein universe\n      <a href="https://www.embopress.org/doi/full/10.1093/emboj/cdg359">\n       [12]\n      </a>\n      . Their high precision and binding affinities have caused their broad adoption in not only therapeutics but commercial and research applications as well as to tag proteins in solution in flow cytometry, CyTOF, immuno-precipitation, and other target identification assays.\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h3 id="1.5-Current-Paradigms-for-Antibodies-as-a-Therapeutic-Modality">\n      <strong>\n       1.5 Current Paradigms for Antibodies as a Therapeutic Modality\n      </strong>\n      <a class="anchor-link" href="#1.5-Current-Paradigms-for-Antibodies-as-a-Therapeutic-Modality">\n       \xb6\n      </a>\n     </h3>\n     <p>\n      Given their unparalleled ability to precisely and durably bind arbitrary targets, there has been a significant interest in possessing antibodies for desired targets. There are a number of therapeutic use cases for these antibodies, for diseases ranging from transplant rejection, non-Hodgkin\'s lymphoma, immune checkpoint inhibitors for cancer immunotherapy, psoriasis, multiple sclerosis, Crohn\'s disease, and many more. While antibodies against common pathogens can be isolated from the serum of convalescent individuals and screened for specificity, the process of procuring novel antibodies is substantially more challenging and involves inoculating an animal with an antigen and isolating the antibodies after. For example,\n      <a href="https://en.wikipedia.org/wiki/Antivenom">\n       anti-venom\n      </a>\n      is a solution of antibodies derived from animals (typically horses) against cytotoxic proteins found in venom. This procedure is both resource and labor intensive. This is because after inoculation and isolation of antibodies from the animal, there is an additional step of screening them for reactivity against the target using reaction chemistry methods such as\n      <a href="https://en.wikipedia.org/wiki/Surface_plasmon_resonance">\n       surface plamon resonance\n      </a>\n      or\n      <a href="https://en.wikipedia.org/wiki/Bio-layer_interferometry">\n       bio-layer interferometry\n      </a>\n      . As such there has been a great deal of interest in methods of\n      <em>\n       in-silico\n      </em>\n      antibody design. A number of approaches have shown reasonable degrees of success in this task from guided evolution based approaches\n      <a href="https://www.nature.com/articles/s41587-023-01763-2">\n       [1]\n      </a>\n      to newer diffusion based\n      <a href="https://arxiv.org/abs/2308.05027">\n       [13]\n      </a>\n      approaches. In this tutorial we will pay homage to the former.\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h2>\n      2. Let\'s Code! Designing Antibodies via Directed Evolution\n      <a class="anchor-link" href="#2.-Let\'s-Code!-Designing-Antibodies-via-Directed-Evolution">\n       \xb6\n      </a>\n     </h2>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h3 id="2.1-Overview">\n      <strong>\n       2.1 Overview\n      </strong>\n      <a class="anchor-link" href="#2.1-Overview">\n       \xb6\n      </a>\n     </h3>\n     <p>\n      Now that we have the minimal backgrounded needed to understand the antibody design proble and the necessary\n      <a href="https://colab.research.google.com/drive/13eXPgZpzTOL3c_S7OM7uR6btP8ZWw7zn#scrollTo=0PMAl8z_HEBr">\n       language model background\n      </a>\n      , we can jump right into antibody design via directed evolution, as shown in the figure below:\n     </p>\n     <p>\n      \\\n     </p>\n     <p>\n      <img alt="directed_evolution" src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41587-023-01991-6/MediaObjects/41587_2023_1991_Fig1_HTML.png?as=webp"/>\n     </p>\n     <p>\n      Image Source: Figure 1.\n      <a href="https://www.nature.com/articles/s41587-023-01991-6">\n       Outeiral et. al\n      </a>\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h3 id="2.2-Setup/Methodology">\n      <strong>\n       2.2 Setup/Methodology\n      </strong>\n      <a class="anchor-link" href="#2.2-Setup/Methodology">\n       \xb6\n      </a>\n     </h3>\n     <p>\n      In Hie et al. the authors decide to use a general protein language model instead of one trained specifically on antibody sequences. They use the ESM-1b and ESM-1v models which were trained on UniRef50 and UniRef90\n      <a href="https://academic.oup.com/bioinformatics/article/23/10/1282/197795?login=true">\n       [14]\n      </a>\n      , respectively. For their directed evolution studies they select seven therapeutic antibodies associated with viral infection spanning Influenza, Ebolavirus, and SARS-COv2. The authors use a straightforward and exhaustive mutation scheduler in mutating every residue in the antigen binding region to every other residue and computing the likelihood of the sequence. Sequences with likelihoods greater than or equal to the WT sequence were kept for experimental validation. For our purposes, we need not be as thorough and can use a slightly expedited method by taking the top-k mutations at a specific point.\n     </p>\n     <p>\n      Inspired by the work of Hie et al., we first define the pLM driven directed evolution task as simply passing in a masked antibody sequence to a pLM that was previously trained on the masked language modeling objective and examining the token probabilities for the masked amino acids. It really is that easy!\n     </p>\n     <p>\n      For reference we break the task down into the following steps:\n     </p>\n     <pre class="overflow-x-scroll font-mono"><code># Antibody Design via pLM Directed Evolution\n1. Select a pre-trained model language model (can be pre-trained on all domains or exclusively antibodies)\n2. Choose an antibody to mutate.\n3. Determine the amino acid(s) to mask out*.\n4. Pass the tokenized sequences into the pLM\n5. Sample tokens according to a heuristic for increased fitness\n</code></pre>\n     <p>\n      *Modification of antibodies needs to focus only on the variable regions as the amino acids at the interface are the ones responsible for driving affinity. Making edits to the constant region would actually be detrimental to antibodies\' effector function in the complement system as well as potentially disrupt binding to innate immune receptors. \\\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h4 id="2.2.1-Loading-the-Model-+-Tokenizer">\n      2.2.1 Loading the Model + Tokenizer\n      <a class="anchor-link" href="#2.2.1-Loading-the-Model-+-Tokenizer">\n       \xb6\n      </a>\n     </h4>\n     <p>\n      For this exploration, we pay homage to an early Antibody Language Model AbLang\n      <a href="https://academic.oup.com/bioinformaticsadvances/article/2/1/vbac046/6609807?login=true">\n       [15]\n      </a>\n      . AbLang is a masked language model based on the RoBERTa\n      <a href="https://arxiv.org/abs/1907.11692">\n       [16]\n      </a>\n      model, and pre-trained on antibody sequences from the [observed antibody space (OAS)\n      <a href="https://opig.stats.ox.ac.uk/webapps/oas/oas_paired/">\n       [17]\n      </a>\n      . AbLang consists of two models, one trained on the heavy chain sequences and one trained on the light chain sequences and the authors demonstrate its usefulness over broader protein language models such as ESM-1b, contradicting the findings put forth in Hie et al. Both the heavy and light chain models are identical in architecture with a $d_{model}$ of 768, max position embedding of 160, and 12 transformer block layers, totaling ~86M parameters.\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-CodeCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n     In\xa0[\xa0]:\n    </div>\n    <div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">\n     <div class="cm-editor cm-s-jupyter">\n      <div class="highlight hl-python">\n       <pre class="overflow-x-scroll font-mono"><span></span><span class="kn">from</span> <span class="nn">transformers</span> <span class="kn">import</span> <span class="n">AutoModel</span><span class="p">,</span> <span class="n">AutoTokenizer</span><span class="p">,</span> <span class="n">AutoModelForMaskedLM</span>\n\n<span class="c1"># Get the tokenizer</span>\n<span class="n">tokenizer</span> <span class="o">=</span> <span class="n">AutoTokenizer</span><span class="o">.</span><span class="n">from_pretrained</span><span class="p">(</span><span class="s1">\'qilowoq/AbLang_light\'</span><span class="p">)</span>\n\n<span class="c1"># Get the light chain model</span>\n<span class="n">mlm_light_chain_model</span> <span class="o">=</span> <span class="n">AutoModelForMaskedLM</span><span class="o">.</span><span class="n">from_pretrained</span><span class="p">(</span><span class="s1">\'qilowoq/AbLang_light\'</span><span class="p">)</span>\n<span class="c1"># Get the heavy chain model</span>\n<span class="n">mlm_heavy_chain_model</span> <span class="o">=</span> <span class="n">AutoModelForMaskedLM</span><span class="o">.</span><span class="n">from_pretrained</span><span class="p">(</span><span class="s1">\'qilowoq/AbLang_heavy\'</span><span class="p">)</span>\n</pre>\n      </div>\n     </div>\n    </div>\n   </div>\n  </div>\n  <div class="jp-Cell-outputWrapper">\n   <div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">\n   </div>\n   <div class="jp-OutputArea jp-Cell-outputArea">\n    <div class="jp-OutputArea-child">\n     <div class="jp-OutputPrompt jp-OutputArea-prompt">\n     </div>\n     <div class="jp-RenderedText jp-OutputArea-output" data-mime-type="application/vnd.jupyter.stderr" tabindex="0">\n      <pre class="overflow-x-scroll font-mono">/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: \nThe secret `HF_TOKEN` does not exist in your Colab secrets.\nTo authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.\nYou will be able to reuse this secret in all of your notebooks.\nPlease note that authentication is recommended but still optional to access public models or datasets.\n  warnings.warn(\n</pre>\n     </div>\n    </div>\n    <div class="jp-OutputArea-child">\n     <div class="jp-OutputPrompt jp-OutputArea-prompt">\n     </div>\n     <div class="jp-RenderedText jp-OutputArea-output" data-mime-type="text/plain" tabindex="0">\n      <pre class="overflow-x-scroll font-mono">tokenizer_config.json:   0%|          | 0.00/367 [00:00&lt;?, ?B/s]</pre>\n     </div>\n    </div>\n    <div class="jp-OutputArea-child">\n     <div class="jp-OutputPrompt jp-OutputArea-prompt">\n     </div>\n     <div class="jp-RenderedText jp-OutputArea-output" data-mime-type="text/plain" tabindex="0">\n      <pre class="overflow-x-scroll font-mono">vocab.txt:   0%|          | 0.00/71.0 [00:00&lt;?, ?B/s]</pre>\n     </div>\n    </div>\n    <div class="jp-OutputArea-child">\n     <div class="jp-OutputPrompt jp-OutputArea-prompt">\n     </div>\n     <div class="jp-RenderedText jp-OutputArea-output" data-mime-type="text/plain" tabindex="0">\n      <pre class="overflow-x-scroll font-mono">tokenizer.json:   0%|          | 0.00/3.02k [00:00&lt;?, ?B/s]</pre>\n     </div>\n    </div>\n    <div class="jp-OutputArea-child">\n     <div class="jp-OutputPrompt jp-OutputArea-prompt">\n     </div>\n     <div class="jp-RenderedText jp-OutputArea-output" data-mime-type="text/plain" tabindex="0">\n      <pre class="overflow-x-scroll font-mono">special_tokens_map.json:   0%|          | 0.00/125 [00:00&lt;?, ?B/s]</pre>\n     </div>\n    </div>\n    <div class="jp-OutputArea-child">\n     <div class="jp-OutputPrompt jp-OutputArea-prompt">\n     </div>\n     <div class="jp-RenderedText jp-OutputArea-output" data-mime-type="text/plain" tabindex="0">\n      <pre class="overflow-x-scroll font-mono">config.json:   0%|          | 0.00/848 [00:00&lt;?, ?B/s]</pre>\n     </div>\n    </div>\n    <div class="jp-OutputArea-child">\n     <div class="jp-OutputPrompt jp-OutputArea-prompt">\n     </div>\n     <div class="jp-RenderedText jp-OutputArea-output" data-mime-type="text/plain" tabindex="0">\n      <pre class="overflow-x-scroll font-mono">pytorch_model.bin:   0%|          | 0.00/343M [00:00&lt;?, ?B/s]</pre>\n     </div>\n    </div>\n    <div class="jp-OutputArea-child">\n     <div class="jp-OutputPrompt jp-OutputArea-prompt">\n     </div>\n     <div class="jp-RenderedText jp-OutputArea-output" data-mime-type="text/plain" tabindex="0">\n      <pre class="overflow-x-scroll font-mono">config.json:   0%|          | 0.00/848 [00:00&lt;?, ?B/s]</pre>\n     </div>\n    </div>\n    <div class="jp-OutputArea-child">\n     <div class="jp-OutputPrompt jp-OutputArea-prompt">\n     </div>\n     <div class="jp-RenderedText jp-OutputArea-output" data-mime-type="text/plain" tabindex="0">\n      <pre class="overflow-x-scroll font-mono">pytorch_model.bin:   0%|          | 0.00/343M [00:00&lt;?, ?B/s]</pre>\n     </div>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-CodeCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n     In\xa0[\xa0]:\n    </div>\n    <div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">\n     <div class="cm-editor cm-s-jupyter">\n      <div class="highlight hl-python">\n       <pre class="overflow-x-scroll font-mono"><span></span><span class="c1"># Lets take a look at the model parameter count and architecture</span>\n<span class="n">n_params</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">numel</span><span class="p">()</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">mlm_heavy_chain_model</span><span class="o">.</span><span class="n">parameters</span><span class="p">())</span>\n<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">\'The Ablang model has </span><span class="si">{</span><span class="n">n_params</span><span class="si">}</span><span class="s1"> trainable parameters. </span><span class="se">\\n</span><span class="s1">\'</span><span class="p">)</span>\n<span class="n">mlm_heavy_chain_model</span>\n</pre>\n      </div>\n     </div>\n    </div>\n   </div>\n  </div>\n  <div class="jp-Cell-outputWrapper">\n   <div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">\n   </div>\n   <div class="jp-OutputArea jp-Cell-outputArea">\n    <div class="jp-OutputArea-child">\n     <div class="jp-OutputPrompt jp-OutputArea-prompt">\n     </div>\n     <div class="jp-RenderedText jp-OutputArea-output" data-mime-type="text/plain" tabindex="0">\n      <pre class="overflow-x-scroll font-mono">The Ablang model has 85809432 trainable parameters. \n\n</pre>\n     </div>\n    </div>\n    <div class="jp-OutputArea-child jp-OutputArea-executeResult">\n     <div class="jp-OutputPrompt jp-OutputArea-prompt">\n      Out[\xa0]:\n     </div>\n     <div class="jp-RenderedText jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/plain" tabindex="0">\n      <pre class="overflow-x-scroll font-mono">RobertaForMaskedLM(\n  (roberta): RobertaModel(\n    (embeddings): RobertaEmbeddings(\n      (word_embeddings): Embedding(24, 768, padding_idx=21)\n      (position_embeddings): Embedding(160, 768, padding_idx=21)\n      (token_type_embeddings): Embedding(2, 768)\n      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n      (dropout): Dropout(p=0.1, inplace=False)\n    )\n    (encoder): RobertaEncoder(\n      (layer): ModuleList(\n        (0-11): 12 x RobertaLayer(\n          (attention): RobertaAttention(\n            (self): RobertaSelfAttention(\n              (query): Linear(in_features=768, out_features=768, bias=True)\n              (key): Linear(in_features=768, out_features=768, bias=True)\n              (value): Linear(in_features=768, out_features=768, bias=True)\n              (dropout): Dropout(p=0.1, inplace=False)\n            )\n            (output): RobertaSelfOutput(\n              (dense): Linear(in_features=768, out_features=768, bias=True)\n              (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n              (dropout): Dropout(p=0.1, inplace=False)\n            )\n          )\n          (intermediate): RobertaIntermediate(\n            (dense): Linear(in_features=768, out_features=3072, bias=True)\n            (intermediate_act_fn): GELUActivation()\n          )\n          (output): RobertaOutput(\n            (dense): Linear(in_features=3072, out_features=768, bias=True)\n            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n            (dropout): Dropout(p=0.1, inplace=False)\n          )\n        )\n      )\n    )\n  )\n  (lm_head): RobertaLMHead(\n    (dense): Linear(in_features=768, out_features=768, bias=True)\n    (layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n    (decoder): Linear(in_features=768, out_features=24, bias=True)\n  )\n)</pre>\n     </div>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h4 id="2.2.2-Example-Antibody-Sequence">\n      2.2.2 Example Antibody Sequence\n      <a class="anchor-link" href="#2.2.2-Example-Antibody-Sequence">\n       \xb6\n      </a>\n     </h4>\n     <p>\n      Next let\'s choose a heavy chain and light chain for designing. These were chosen from the Ablang examples page on HuggingFace,\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-CodeCell jp-Notebook-cell jp-mod-noOutputs">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n     In\xa0[\xa0]:\n    </div>\n    <div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">\n     <div class="cm-editor cm-s-jupyter">\n      <div class="highlight hl-python">\n       <pre class="overflow-x-scroll font-mono"><span></span><span class="c1"># Lets take the variable regions of the heavy and light chains</span>\n\n<span class="n">heavy_chain_example</span> <span class="o">=</span>  <span class="s1">\'EVQLQESGPGLVKPSETLSLTCTVSGGPINNAYWTWIRQPPGKGLEYLGYVYHTGVTNYNPSLKSRLTITIDTSRKQLSLSLKFVTAADSAVYYCAREWAEDGDFGNAFHVWGQGTMVAVSSASTKGPSVFPLAPSSKSTSGGTAALGCL\'</span>\n<span class="n">light_chain_example</span> <span class="o">=</span> <span class="s1">\'GSELTQDPAVSVALGQTVRITCQGDSLRNYYASWYQQKPRQAPVLVFYGKNNRPSGIPDRFSGSSSGNTASLTISGAQAEDEADYYCNSRDSSSNHLVFGGGTKLTVLSQ\'</span>\n</pre>\n      </div>\n     </div>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h4 id="2.2.3-Masking-the-Sequence">\n      2.2.3 Masking the Sequence\n      <a class="anchor-link" href="#2.2.3-Masking-the-Sequence">\n       \xb6\n      </a>\n     </h4>\n     <p>\n      One of the crucial parameters with this approach is in the determination of which residues to mask and re-design. Let\'s start off by first setting up some reproducible code so that we can apply the masking procedure to any number of sequences at arbitrary points.\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-CodeCell jp-Notebook-cell jp-mod-noOutputs">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n     In\xa0[\xa0]:\n    </div>\n    <div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">\n     <div class="cm-editor cm-s-jupyter">\n      <div class="highlight hl-python">\n       <pre class="overflow-x-scroll font-mono"><span></span><span class="c1"># Sequnece masking convenience function</span>\n<span class="k">def</span> <span class="nf">mask_seq_pos</span><span class="p">(</span><span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>\n                      <span class="n">idx</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>\n                      <span class="n">mask</span><span class="o">=</span><span class="s1">\'[MASK]\'</span><span class="p">):</span>\n    <span class="sd">\'\'\'Given an arbitrary antibody sequence with and a seqeunce index,</span>\n<span class="sd">    convert the residue at that index into the mask token.</span>\n\n<span class="sd">    \'\'\'</span>\n    <span class="n">cleaned_sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">\' \'</span><span class="p">,</span> <span class="s1">\'\'</span><span class="p">)</span>  <span class="c1"># Get ride of extraneous spaces if any</span>\n    <span class="k">assert</span> <span class="nb">abs</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="s2">"Zero-indexed value needs to be less than sequence length minus one."</span>\n    <span class="n">cleaned_sequence</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">cleaned_sequence</span><span class="p">)</span>  <span class="c1"># Turn the sequence into a list</span>\n    <span class="n">cleaned_sequence</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="s1">\'*\'</span>                 <span class="c1"># Mask the sequence at idx</span>\n    <span class="n">masked_sequence</span> <span class="o">=</span> <span class="s1">\' \'</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cleaned_sequence</span><span class="p">)</span> <span class="c1"># Convert list -&gt; seq</span>\n    <span class="n">masked_sequence</span> <span class="o">=</span> <span class="n">masked_sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">\'*\'</span><span class="p">,</span> <span class="n">mask</span><span class="p">)</span>\n    <span class="k">return</span> <span class="n">masked_sequence</span>\n\n<span class="c1"># Test</span>\n<span class="k">assert</span> <span class="n">mask_seq_pos</span><span class="p">(</span><span class="s1">\'CAT\'</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span><span class="o">==</span><span class="s1">\'C [MASK] T\'</span>\n\n<span class="c1">#TODO: Add unit tests with pytest where you can check that the assert has been hit</span>\n</pre>\n      </div>\n     </div>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h4 id="2.2.3-Model-Inference">\n      2.2.3 Model Inference\n      <a class="anchor-link" href="#2.2.3-Model-Inference">\n       \xb6\n      </a>\n     </h4>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-CodeCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n     In\xa0[\xa0]:\n    </div>\n    <div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">\n     <div class="cm-editor cm-s-jupyter">\n      <div class="highlight hl-python">\n       <pre class="overflow-x-scroll font-mono"><span></span><span class="c1">### Step 1. Mask the light_chain sequence</span>\n<span class="n">mask_idx</span> <span class="o">=</span> <span class="mi">9</span>\n<span class="n">masked_light_chain</span> <span class="o">=</span> <span class="n">mask_seq_pos</span><span class="p">(</span><span class="n">light_chain_example</span><span class="p">,</span> <span class="n">idx</span><span class="o">=</span><span class="n">mask_idx</span><span class="p">)</span>\n<span class="c1">### Step 2. Tokenize</span>\n<span class="n">tokenized_input</span> <span class="o">=</span> <span class="n">tokenizer</span><span class="p">(</span><span class="n">masked_light_chain</span><span class="p">,</span> <span class="n">return_tensors</span><span class="o">=</span><span class="s1">\'pt\'</span><span class="p">)</span>\n<span class="c1">### Step 3. Light Chain Model</span>\n<span class="n">mlm_output</span> <span class="o">=</span> <span class="n">mlm_light_chain_model</span><span class="p">(</span><span class="o">**</span><span class="n">tokenized_input</span><span class="p">)</span>\n<span class="c1">### Step 4. Decode the outputs to see what the model has placed</span>\n<span class="n">decoded_outs</span> <span class="o">=</span> <span class="n">tokenizer</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="n">mlm_output</span><span class="o">.</span><span class="n">logits</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">skip_special_tokens</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>\n<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">\'Model predicted: </span><span class="si">{</span><span class="n">decoded_outs</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">" "</span><span class="p">,</span> <span class="s2">""</span><span class="p">)[</span><span class="mi">9</span><span class="p">]</span><span class="si">}</span><span class="s1"> at index </span><span class="si">{</span><span class="n">mask_idx</span><span class="si">}</span><span class="s1">\'</span><span class="p">)</span>\n<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">\'Predicted Sequence: </span><span class="si">{</span><span class="n">decoded_outs</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">" "</span><span class="p">,</span> <span class="s2">""</span><span class="p">)</span><span class="si">}</span><span class="s1">\'</span><span class="p">)</span>\n<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">\'Starting  Sequence: </span><span class="si">{</span><span class="n">light_chain_example</span><span class="si">}</span><span class="s1">\'</span><span class="p">)</span>\n</pre>\n      </div>\n     </div>\n    </div>\n   </div>\n  </div>\n  <div class="jp-Cell-outputWrapper">\n   <div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">\n   </div>\n   <div class="jp-OutputArea jp-Cell-outputArea">\n    <div class="jp-OutputArea-child">\n     <div class="jp-OutputPrompt jp-OutputArea-prompt">\n     </div>\n     <div class="jp-RenderedText jp-OutputArea-output" data-mime-type="text/plain" tabindex="0">\n      <pre class="overflow-x-scroll font-mono">Model predicted: S at index 9\nPredicted Sequence: SADSSSCGVSSTVAHGQTLKINSQGQRHSLYYVRWYQQKPGLAPLLLIYGKNSRPSGIPDRFSGSKSGTTASLTITGLQAEDEADYYCQQSGGSGGHLTVGGGALLATLTQ\nStarting  Sequence: GSELTQDPAVSVALGQTVRITCQGDSLRNYYASWYQQKPRQAPVLVFYGKNNRPSGIPDRFSGSSSGNTASLTISGAQAEDEADYYCNSRDSSSNHLVFGGGTKLTVLSQ\n</pre>\n     </div>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h3 id="2.2.4-HuggingFace-Pipeline-Object">\n      2.2.4 HuggingFace Pipeline Object\n      <a class="anchor-link" href="#2.2.4-HuggingFace-Pipeline-Object">\n       \xb6\n      </a>\n     </h3>\n     <p>\n      Hold on, given the tokenized input with only one masked token, we would expect to see only one change to the the sequence. However, what we get back is something a lot more different that what we put in. Luckily, there\'s something in the HuggingFace software suite that we can use to address this: Pipelines\n     </p>\n     <p>\n      HuggingFace Pipelines:\n     </p>\n     <ol>\n      <li>\n       Pipeline object is a wrapper for inference and can be treated like an object for API calls\n      </li>\n      <li>\n       There is a fill-mask pipeline that we can use which accepts a single mask token in out input and outputs a dictionary of the score of that sequence, the imputed token, and the reconstructed full sequence.\n      </li>\n     </ol>\n     <p>\n      Lets see it in action:\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-CodeCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n     In\xa0[\xa0]:\n    </div>\n    <div class="jp-CodeMirrorEditor jp-Editor jp-InputArea-editor" data-type="inline">\n     <div class="cm-editor cm-s-jupyter">\n      <div class="highlight hl-python">\n       <pre class="overflow-x-scroll font-mono"><span></span><span class="kn">from</span> <span class="nn">transformers</span> <span class="kn">import</span> <span class="n">pipeline</span>\n<span class="n">filler</span> <span class="o">=</span> <span class="n">pipeline</span><span class="p">(</span><span class="n">task</span><span class="o">=</span><span class="s1">\'fill-mask\'</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">mlm_light_chain_model</span><span class="p">,</span> <span class="n">tokenizer</span><span class="o">=</span><span class="n">tokenizer</span><span class="p">)</span>\n<span class="n">filler</span><span class="p">(</span><span class="n">masked_light_chain</span><span class="p">)</span> <span class="c1"># fill in the mask</span>\n</pre>\n      </div>\n     </div>\n    </div>\n   </div>\n  </div>\n  <div class="jp-Cell-outputWrapper">\n   <div class="jp-Collapser jp-OutputCollapser jp-Cell-outputCollapser">\n   </div>\n   <div class="jp-OutputArea jp-Cell-outputArea">\n    <div class="jp-OutputArea-child jp-OutputArea-executeResult">\n     <div class="jp-OutputPrompt jp-OutputArea-prompt">\n      Out[\xa0]:\n     </div>\n     <div class="jp-RenderedText jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/plain" tabindex="0">\n      <pre class="overflow-x-scroll font-mono">[{\'score\': 0.13761496543884277,\n  \'token\': 7,\n  \'token_str\': \'S\',\n  \'sequence\': \'G S E L T Q D P A S S V A L G Q T V R I T C Q G D S L R N Y Y A S W Y Q Q K P R Q A P V L V F Y G K N N R P S G I P D R F S G S S S G N T A S L T I S G A Q A E D E A D Y Y C N S R D S S S N H L V F G G G T K L T V L S Q\'},\n {\'score\': 0.1152879148721695,\n  \'token\': 6,\n  \'token_str\': \'E\',\n  \'sequence\': \'G S E L T Q D P A E S V A L G Q T V R I T C Q G D S L R N Y Y A S W Y Q Q K P R Q A P V L V F Y G K N N R P S G I P D R F S G S S S G N T A S L T I S G A Q A E D E A D Y Y C N S R D S S S N H L V F G G G T K L T V L S Q\'},\n {\'score\': 0.0989701896905899,\n  \'token\': 9,\n  \'token_str\': \'N\',\n  \'sequence\': \'G S E L T Q D P A N S V A L G Q T V R I T C Q G D S L R N Y Y A S W Y Q Q K P R Q A P V L V F Y G K N N R P S G I P D R F S G S S S G N T A S L T I S G A Q A E D E A D Y Y C N S R D S S S N H L V F G G G T K L T V L S Q\'},\n {\'score\': 0.08586061000823975,\n  \'token\': 14,\n  \'token_str\': \'A\',\n  \'sequence\': \'G S E L T Q D P A A S V A L G Q T V R I T C Q G D S L R N Y Y A S W Y Q Q K P R Q A P V L V F Y G K N N R P S G I P D R F S G S S S G N T A S L T I S G A Q A E D E A D Y Y C N S R D S S S N H L V F G G G T K L T V L S Q\'},\n {\'score\': 0.07652082294225693,\n  \'token\': 8,\n  \'token_str\': \'T\',\n  \'sequence\': \'G S E L T Q D P A T S V A L G Q T V R I T C Q G D S L R N Y Y A S W Y Q Q K P R Q A P V L V F Y G K N N R P S G I P D R F S G S S S G N T A S L T I S G A Q A E D E A D Y Y C N S R D S S S N H L V F G G G T K L T V L S Q\'}]</pre>\n     </div>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <p>\n      <strong>\n       Congratulations you have now designed 5 new antibodies!\n      </strong>\n     </p>\n     <p>\n      Disclaimer: For a more thorough antibody (re)design, we will typically want to follow an approach like what was done in Hie et al. where every point along the sequence will be mutated and the total number of sequences will be collated and scored with the top-100 or so antibodies being expressed for validation. If you would like to explore this feel free to try it out yourself as a challenge!\n     </p>\n     <p>\n      You can also refer to the real data in Hie et al. to see if any of the predicted ones were found to work well and increase fitness.\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h3 id="2.3-Limitations">\n      <strong>\n       2.3 Limitations\n      </strong>\n      <a class="anchor-link" href="#2.3-Limitations">\n       \xb6\n      </a>\n     </h3>\n     <p>\n      While promising, this approach is obviously not without its shortcomings. Key limitations include:\n     </p>\n     <ul>\n      <li>\n       Fixed length antibody design since masked tokens are applied in a 1:1 fashion.\n      </li>\n      <li>\n       Lack of target information included during conditional sampling step which can influence choice of amino acid given the sequence context.\n      </li>\n      <li>\n       Approach is sensitive to choice of protein language model\n      </li>\n     </ul>\n     <p>\n      This letter\n      <a href="https://www.nature.com/articles/s41587-023-01991-6">\n       [18]\n      </a>\n      provides a great synopsis of Hie et al.\'s work, which by extension apply to the methods presented in this tutorial as well.\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h2>\n      Citing this Tutorial\n      <a class="anchor-link" href="#Citing-this-Tutorial">\n       \xb6\n      </a>\n     </h2>\n     <p>\n      If you found this tutorial useful, please consider citing it as:\n     </p>\n     <pre class="overflow-x-scroll font-mono"><code>@manual{Bioinformatics,\n title={An Introduction to Antibody Design Using Protein Language Models},\n organization={DeepChem},\n author={Karthikeyan, Dhuvarakesh and Menezes, Aaron},\n howpublished = {\\url{https://github.com/deepchem/deepchem/blob/master/examples/tutorials/DeepChem_AntibodyTutorial_Simplified.ipynb}},\n year={2024},\n}\n</code></pre>\n    </div>\n   </div>\n  </div>\n </div>\n <div class="jp-Cell jp-MarkdownCell jp-Notebook-cell">\n  <div class="jp-Cell-inputWrapper" tabindex="0">\n   <div class="jp-Collapser jp-InputCollapser jp-Cell-inputCollapser">\n   </div>\n   <div class="jp-InputArea jp-Cell-inputArea">\n    <div class="jp-InputPrompt jp-InputArea-prompt">\n    </div>\n    <div class="jp-RenderedHTMLCommon jp-RenderedMarkdown jp-MarkdownOutput" data-mime-type="text/markdown">\n     <h2>\n      Works Cited\n      <a class="anchor-link" href="#Works-Cited">\n       \xb6\n      </a>\n     </h2>\n     <p>\n      [1] Hie, B.L., Shanker, V.R., Xu, D. et al. Efficient evolution of human antibodies from general protein language models. Nat Biotechnol 42, 275–283 (2024).\n      <a href="https://doi.org/10.1038/s41587-023-01763-2">\n       https://doi.org/10.1038/s41587-023-01763-2\n      </a>\n     </p>\n     <p>\n      [2] Bretscher, P., &amp; Cohn, M. (1970). A Theory of Self-Nonself Discrimination. Science, 169(3950), 1042–1049. doi:10.1126/science.169.3950.1042\n     </p>\n     <p>\n      [3] Cohn, M. The common sense of the self-nonself discrimination. Springer Semin Immun 27, 3–17 (2005).\n      <a href="https://doi.org/10.1007/s00281-005-0199-1">\n       https://doi.org/10.1007/s00281-005-0199-1\n      </a>\n     </p>\n     <p>\n      [4] Gonzalez S, Gonz\xe1lez-Rodr\xedguez AP, Su\xe1rez-\xc1lvarez B, L\xf3pez-Soto A, Huergo-Zapico L, Lopez-Larrea C. Conceptual aspects of self and nonself discrimination. Self Nonself. 2011 Jan;2(1):19-25. doi: 10.4161/self.2.1.15094. Epub 2011 Jan 1. PMID: 21776331; PMCID: PMC3136900.\n     </p>\n     <p>\n      [5] ROB J. DE BOER, PAULINE HOGEWEG, Self-Nonself Discrimination due to Immunological Nonlinearities: the Analysis of a Series of Models by Numerical Methods, Mathematical Medicine and Biology: A Journal of the IMA, Volume 4, Issue 1, 1987, Pages 1–32,\n      <a href="https://doi.org/10.1093/imammb/4.1.1">\n       https://doi.org/10.1093/imammb/4.1.1\n      </a>\n     </p>\n     <p>\n      [6] Cohn, M. A biological context for the self-nonself discrimination and the regulation of effector class by the immune system. Immunol Res 31, 133–150 (2005).\n      <a href="https://doi.org/10.1385/IR:31:2:133">\n       https://doi.org/10.1385/IR:31:2:133\n      </a>\n     </p>\n     <p>\n      [7] Janeway CA Jr, Travers P, Walport M, et al. Immunobiology: The Immune System in Health and Disease. 5th edition. New York: Garland Science; 2001. Principles of innate and adaptive immunity. Available from:\n      <a href="https://www.ncbi.nlm.nih.gov/books/NBK27090/">\n       https://www.ncbi.nlm.nih.gov/books/NBK27090/\n      </a>\n     </p>\n     <p>\n      [8] Perelson, A. Modelling viral and immune system dynamics. Nat Rev Immunol. 2. , 28–36 (2002).\n      <a href="https://doi.org/10.1038/nri700">\n       https://doi.org/10.1038/nri700\n      </a>\n     </p>\n     <p>\n      [9] Shittu, A. (n.d.). Understanding immunological memory. ASM.org.\n      <a href="https://asm.org/articles/2023/may/understanding-immunological-memory">\n       https://asm.org/articles/2023/may/understanding-immunological-memory\n      </a>\n     </p>\n     <p>\n      [10] Janeway CA Jr, Travers P, Walport M, et al. Immunobiology: The Immune System in Health and Disease. 5th edition. New York: Garland Science; 2001. Chapter 8, T Cell-Mediated Immunity. Available from:\n      <a href="https://www.ncbi.nlm.nih.gov/books/NBK10762/">\n       https://www.ncbi.nlm.nih.gov/books/NBK10762/\n      </a>\n     </p>\n     <p>\n      [11] Glick, B., Chang, T. S., &amp; Jaap, R. G. (1956). The Bursa of Fabricius and Antibody Production. Poultry Science, 35(1), 224–225. doi:10.3382/ps.0350224\n     </p>\n     <p>\n      [12] Nooren, I. M. (2003). NEW EMBO MEMBER’S REVIEW: Diversity of protein-protein interactions. EMBO Journal, 22(14), 3486–3492.\n      <a href="https://doi.org/10.1093/emboj/cdg359">\n       https://doi.org/10.1093/emboj/cdg359\n      </a>\n     </p>\n     <p>\n      [13] Karolis Martinkus, Jan Ludwiczak, Kyunghyun Cho, Wei-Ching Liang, Julien Lafrance-Vanasse, Isidro Hotzel, Arvind Rajpal, Yan Wu, Richard Bonneau, Vladimir Gligorĳevic, &amp; Andreas Loukas. (2024). AbDiffuser: Full-Atom Generation of in vitro Functioning Antibodies.\n     </p>\n     <p>\n      [14] Baris E. Suzek, Hongzhan Huang, Peter McGarvey, Raja Mazumder, Cathy H. Wu, UniRef: comprehensive and non-redundant UniProt reference clusters, Bioinformatics, Volume 23, Issue 10, May 2007, Pages 1282–1288,\n      <a href="https://doi.org/10.1093/bioinformatics/btm098">\n       https://doi.org/10.1093/bioinformatics/btm098\n      </a>\n     </p>\n     <p>\n      [15] Tobias H Olsen, Iain H Moal, Charlotte M Deane, AbLang: an antibody language model for completing antibody sequences, Bioinformatics Advances, Volume 2, Issue 1, 2022, vbac046,\n      <a href="https://doi.org/10.1093/bioadv/vbac046">\n       https://doi.org/10.1093/bioadv/vbac046\n      </a>\n     </p>\n     <p>\n      [16] Yinhan Liu, Myle Ott, Naman Goyal, Jingfei Du, Mandar Joshi, Danqi Chen, Omer Levy, Mike Lewis, Luke Zettlemoyer, &amp; Veselin Stoyanov. (2019). RoBERTa: A Robustly Optimized BERT Pretraining Approach.\n     </p>\n     <p>\n      [17] Olsen TH, Boyles F, Deane CM. Observed Antibody Space: A diverse database of cleaned, annotated, and translated unpaired and paired antibody sequences. Protein Sci. 2022 Jan;31(1):141-146. doi: 10.1002/pro.4205. Epub 2021 Oct 29. PMID: 34655133; PMCID: PMC8740823.\n     </p>\n     <p>\n      [18] Outeiral, C., Deane, C.M. Perfecting antibodies with language models.\nNat Biotechnol 42, 185–186 (2024).\n      <a href="https://doi.org/10.1038/s41587-023-01991-6">\n       https://doi.org/10.1038/s41587-023-01991-6\n      </a>\n     </p>\n    </div>\n   </div>\n  </div>\n </div>\n</main>\n'},p=a(7294),l=a(7466),r=a.n(l);let d=()=>((0,p.useEffect)(()=>{var n,e;null===(n=document.getElementsByClassName("scroll-nav")[0])||void 0===n||n.remove();let a=document.querySelector(".notebook"),s=document.querySelector(".notebook");s&&a&&r().init(a,{sections:"h1, h2",insertTarget:s,insertLocation:"after"}),null==MathJax||null===(e=MathJax.Hub)||void 0===e||e.Queue(["Typeset",MathJax.Hub])},[]),(0,s.jsx)("div",{className:"overflow-x-scroll",dangerouslySetInnerHTML:{__html:"".concat(o.html," ").concat(i.Z)}}));d.Layout=t.Z;var c=d}},function(n){n.O(0,[2443,9774,2888,179],function(){return n(n.s=9480)}),_N_E=n.O()}]);