(self.webpackChunk_N_E=self.webpackChunk_N_E||[]).push([[8204],{5669:function(e,i,t){(window.__NEXT_P=window.__NEXT_P||[]).push(["/tutorials",function(){return t(4591)}])},1514:function(e,i){"use strict";i.Z=[{title:"The Basic Tools of the Deep Life Sciences",urlifiedFileName:"the-basic-tools-of-the-deep-life-sciences"},{title:"Working With Datasets",urlifiedFileName:"working-with-datasets"},{title:"An Introduction To MoleculeNet",urlifiedFileName:"an-introduction-to-moleculenet"},{title:"Molecular Fingerprints",urlifiedFileName:"molecular-fingerprints"},{title:"Creating Models with TensorFlow and PyTorch",urlifiedFileName:"creating-models-with-tensorflow-and-pytorch"},{title:"Introduction to Graph Convolutions",urlifiedFileName:"introduction-to-graph-convolutions"},{title:"Going Deeper on Molecular Featurizations",urlifiedFileName:"going-deeper-on-molecular-featurizations"},{title:"Working With Splitters",urlifiedFileName:"working-with-splitters"},{title:"Advanced Model Training",urlifiedFileName:"advanced-model-training"},{title:"Creating a high fidelity model from experimental data",urlifiedFileName:"creating-a-high-fidelity-model-from-experimental-data"},{title:"Putting Multitask Learning to Work",urlifiedFileName:"putting-multitask-learning-to-work"},{title:"Modeling Protein Ligand Interactions",urlifiedFileName:"modeling-protein-ligand-interactions"},{title:"Modeling Protein Ligand Interactions With Atomic Convolutions",urlifiedFileName:"modeling-protein-ligand-interactions-with-atomic-convolutions"},{title:"Conditional Generative Adversarial Networks",urlifiedFileName:"conditional-generative-adversarial-networks"},{title:"Training a Generative Adversarial Network on MNIST",urlifiedFileName:"training-a-generative-adversarial-network-on-mnist"},{title:"Advanced model training using hyperopt",urlifiedFileName:"advanced-model-training-using-hyperopt"},{title:"Introduction to Gaussian Processes",urlifiedFileName:"introduction-to-gaussian-processes"},{title:"PytorchLightning Integration",urlifiedFileName:"pytorchlightning-integration"},{title:"Molecular Fingerprints",urlifiedFileName:"molecular-fingerprints"},{title:"Going Deeper on Molecular Featurizations",urlifiedFileName:"going-deeper-on-molecular-featurizations"},{title:"Learning Unsupervised Embeddings for Molecules",urlifiedFileName:"learning-unsupervised-embeddings-for-molecules"},{title:"Synthetic Feasibility Scoring",urlifiedFileName:"synthetic-feasibility-scoring"},{title:"Atomic Contributions for Molecules",urlifiedFileName:"atomic-contributions-for-molecules"},{title:"Interactive Model Evaluation with Trident Chemwidgets",urlifiedFileName:"interactive-model-evaluation-with-trident-chemwidgets"},{title:"Transfer Learning With ChemBERTa Transformers",urlifiedFileName:"transfer-learning-with-chemberta-transformers"},{title:"Training a Normalizing Flow on QM9",urlifiedFileName:"training-a-normalizing-flow-on-qm9"},{title:"Large Scale Chemical Screens",urlifiedFileName:"large-scale-chemical-screens"},{title:"Introduction to Molecular Attention Transformer",urlifiedFileName:"introduction-to-molecular-attention-transformer"},{title:"Generating molecules with MolGAN",urlifiedFileName:"generating-molecules-with-molgan"},{title:"Introduction to GROVER",urlifiedFileName:"introduction-to-grover"},{title:"Protein Deep Learning",urlifiedFileName:"protein-deep-learning"},{title:"Modeling Protein Ligand Interactions",urlifiedFileName:"modeling-protein-ligand-interactions"},{title:"Modeling Protein Ligand Interactions With Atomic Convolutions",urlifiedFileName:"modeling-protein-ligand-interactions-with-atomic-convolutions"},{title:"DeepChemXAlphafold",urlifiedFileName:"deepchemxalphafold"},{title:"Exploring Quantum Chemistry with GDB1k",urlifiedFileName:"exploring-quantum-chemistry-with-gdb1k"},{title:"DeepQMC tutorial",urlifiedFileName:"deepqmc-tutorial"},{title:" Training an Exchange Correlation Functional using Deepchem",urlifiedFileName:"training-an-exchange-correlation-functional-using-deepchem"},{title:"Introduction to Bioinformatics",urlifiedFileName:"introduction-to-bioinformatics"},{title:"Multisequence Alignments",urlifiedFileName:"multisequence-alignments"},{title:"Scanpy",urlifiedFileName:"scanpy"},{title:"Deep probabilistic analysis of single-cell omics data",urlifiedFileName:"deep-probabilistic-analysis-of-single-cell-omics-data"},{title:"Cell Counting Tutorial",urlifiedFileName:"cell-counting-tutorial"},{title:"Introduction To Material Science",urlifiedFileName:"introduction-to-material-science"},{title:"Using Reinforcement Learning to Play Pong",urlifiedFileName:"using-reinforcement-learning-to-play-pong"},{title:"Introduction to Model Interpretability",urlifiedFileName:"introduction-to-model-interpretability"},{title:"Uncertainty In Deep Learning",urlifiedFileName:"uncertainty-in-deep-learning"},{title:"Physics Informed Neural Networks",urlifiedFileName:"physics-informed-neural-networks"},{title:"Introducing JaxModel and PINNModel",urlifiedFileName:"introducing-jaxmodel-and-pinnmodel"},{title:"About nODE Using Torchdiffeq in Deepchem",urlifiedFileName:"about-node-using-torchdiffeq-in-deepchem"},{title:"Differentiation Infrastructure in Deepchem",urlifiedFileName:"differentiation-infrastructure-in-deepchem"},{title:"Introduction to Equivariance",urlifiedFileName:"introduction-to-equivariance"},{title:" Predict Multi Label Odor Descriptors using OpenPOM",urlifiedFileName:"predict-multi-label-odor-descriptors-using-openpom"},{title:" An Introduction to the Polymers and Their Representation",urlifiedFileName:"an-introduction-to-the-polymers-and-their-representation"}]},4591:function(e,i,t){"use strict";t.r(i);var n=t(5893),l=t(1514),r=t(1163),o=t(7294);let a=()=>{let e=(0,r.useRouter)(),i=l.Z[0];return(0,o.useEffect)(()=>{e.replace("tutorials/".concat(i.urlifiedFileName))},[]),(0,n.jsx)(n.Fragment,{})};i.default=a},1163:function(e,i,t){e.exports=t(880)}},function(e){e.O(0,[9774,2888,179],function(){return e(e.s=5669)}),_N_E=e.O()}]);